soweego/validator/enrichment.py

```diff
@@ -68,56 +83,82 @@ def works_people_cli(catalog, entity, upload, sandbox, dir_io):
+                to_upload.add(stmt)
 
     if upload:
         wikidata_bot.add_works_statements(to_upload, catalog, sandbox)
 
 
-def generate_statements(catalog, entity, bucket_size=5000):
+def generate_statements(catalog: str, entity: str, bucket_size: int = 5000) -> Iterator[Tuple]:
+    """Generate statements about works by people.
+
+    **How it works:**
+
+    1. gather works and people identifiers of the given catalog
+       from relevant Wikidata items
+    2. leverage catalog relationships between works and people
+    3. build Wikidata statements accordingly
+
+    :param catalog: ``{'discogs', 'imdb', 'musicbrainz'}``.
+      A supported catalog
+    :param entity: ``{'actor', 'band', 'director', 'musician', 'producer',
+      'writer', 'audiovisual_work', 'musical_work'}``.
+      A supported entity
+    :param bucket_size: (optional) how many target IDs should be looked up
+      in the given catalog. For efficiency purposes
+    :return: the statements ``generator``,
+      yielding *(work_QID, PID, person_QID, person_catalog_ID)* ``tuple`` s
+    """
     works, people = {}, {}
-    claim_pid = vocabulary.WORKS_BY_PEOPLE_MAPPING[catalog][entity]
-    # Gather works IDs
-    data_gathering.gather_target_ids(
-        target_database.get_work_type(catalog, entity),
-        catalog,
-        target_database.get_work_pid(catalog),
-        works,
-    )
-    # Gather people IDs
-    data_gathering.gather_target_ids(
-        entity, catalog, target_database.get_person_pid(catalog), people
-    )
+
+    # Wikidata side
+    _gather_wd_data(catalog, entity, works, people)
+
     # Invert & simplify dictionaries for easier lookup later on
     works_inverted, people_inverted = (
         _invert_and_simplify(works),
         _invert_and_simplify(people),
     )
-    del works, people
-    # Make buckets: more queries, but more efficient ones
-    works_buckets = utils.make_buckets(
-        list(works_inverted.keys()), bucket_size=bucket_size
+    del works, people  # Efficiency paranoia
+
+    # Make buckets for target queries:
+    # more queries, but more efficient ones
+    total_queries, works_buckets, people_buckets = _prepare_target_queries(
+        bucket_size,
+        works_inverted,
+        people_inverted
     )
-    people_buckets = utils.make_buckets(
-        list(people_inverted.keys()), bucket_size=bucket_size
-    )
-    total_queries = len(works_buckets) * len(people_buckets)
 
+    # Target side
     LOGGER.info(
         'Firing %d queries to the internal database, this will take a while ...',
         total_queries,
     )
 
+    yield from _gather_target_data(
+        catalog, entity, total_queries, works_buckets,
+        works_inverted, people_buckets, people_inverted
+    )
+
+    LOGGER.info('Queries done, statements generated')
+
+
+def _gather_target_data(catalog, entity, total_queries, works_buckets,
+                        works_inverted, people_buckets, people_inverted):
```

Comment:
Why these inverted variables are not generated directly inside the function? Are they used in many places?

Reply:
Yes, they are also used in `_prepare_target_queries` .